### PR TITLE
damldocs: Only strip implied typeclass member constraint, not all constraints.

### PR DESCRIPTION
**Pull request by associahedron**
I realized the logic used for stripping out the implied context in typeclass member function docs was overzealous, causing all constraints to be removed. So this PR fixes that and adds a test case.

This is such an edge case that I'm not surprised it wasn't noticed before.

### CODE REVIEW

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs [hunk 1]
```diff
@@ -330,6 +329,34 @@ getClsDocs ctx@DocCtx{..} (DeclData (L _ (TyClD _ c@ClassDecl{..})) docs) = do
...
+        { fct_context = do
+            context <- fct_context fn
+            case context of
+                TypeTuple xs -> do
+                    let xs' = filter (not . matchesMemberContext cl_anchor cl_args) xs
+                    guard (notNull xs')
+                    Just (TypeTuple xs')
+
+                _ -> Just context
+                    -- TODO: Move to using a more appropriate type
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs]: Do we expect to ever reach this case atm? If not, maybe just call `error` so we see if we mess it up instead of silently doing something that might be wrong?

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs [hunk 2]
```diff
@@ -330,6 +329,34 @@ getClsDocs ctx@DocCtx{..} (DeclData (L _ (TyClD _ c@ClassDecl{..})) docs) = do
...
+                    -- for contexts in damldocs, to avoid this case.
+        }
+
+    -- | Is this the implied context for member functions? We use an anchor
+    -- for comparison because it is more accurate than typenames (which are
+    -- generally susceptible to qualification and shadowing).
+    matchesMemberContext :: Maybe Anchor -> [T.Text] -> DDoc.Type -> Bool
+    matchesMemberContext cl_anchor cl_args ty = and
+        [ cl_anchor == getTypeAppAnchor ty
+        , Just [(Just (Typename arg), Just []) | arg <- cl_args] ==
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs]: I guess we can’t use an anchor here as well since there is no easy way to get access to one?

File: compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{-# LANGUAGE ConstrainedClassMethods #-}
+
+daml 1.2
+-- | This module tests the case where a class method contains a constraint
+-- not present in the class itself.
+module ConstrainedClassMethod where
+
+class A t where
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml]: fwiw it’s not that much of an edge case `foldMap` is the first non-artifical example that comes to mind.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs (hunk 2 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs]: These are type variables so there isn't an anchor.

Comment (cocreature) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs]: Makes sense, thanks for the explanation.

**Comment by cocreature** [comment]
CI failures should go away after rebasing.

File: compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml (hunk 3 above)
Comment (associahedron) [reply compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml]: Good point! 👍
